node/src/reactor/main_reactor/keep_up.rs

```diff
@@ -0,0 +1,422 @@
+        info!(
+            "Historical: sync leap received for: {:?}",
+            best_available.trusted_block_header.block_hash()
+        );
+        let era_validator_weights =
+            best_available.era_validator_weights(self.validator_matrix.fault_tolerance_threshold());
+        for evw in era_validator_weights {
+            let era_id = evw.era_id();
+            if self.validator_matrix.register_era_validator_weights(evw) {
+                info!("Historical: got era: {}", era_id);
+            } else {
+                debug!("Historical: already had era: {}", era_id);
+            }
+        }
+        KeepUpInstruction::CheckLater("Historical: sync leap received".to_string(), Duration::ZERO)
+    }
+
+    fn sync_back_register(
+        &mut self,
+        effect_builder: EffectBuilder<MainEvent>,
+        rng: &mut NodeRng,
+        parent_hash: BlockHash,
+    ) -> KeepUpInstruction {
+        if self.block_synchronizer.register_block_by_hash(
+            parent_hash,
+            true,
+            true,
+            self.chainspec.core_config.simultaneous_peer_requests,
+        ) {
+            info!("Historical: register_block_by_hash: {}", parent_hash);
+            let peers_to_ask = self.net.fully_connected_peers_random(
+                rng,
+                self.chainspec.core_config.simultaneous_peer_requests as usize,
+            );
+            debug!("Historical: peers count: {:?}", peers_to_ask.len());
+            self.block_synchronizer
+                .register_peers(parent_hash, peers_to_ask);
+
+            KeepUpInstruction::Do(
+                Duration::ZERO,
+                effect_builder.immediately().event(|_| {
+                    MainEvent::BlockSynchronizerRequest(BlockSynchronizerRequest::NeedNext)
+                }),
+            )
+        } else {
+            self.block_synchronizer.purge_historical();
+            KeepUpInstruction::CheckLater(
+                "Historical: purged".to_string(),
+                self.control_logic_default_delay.into(),
+            )
+        }
+    }
+
+    fn sync_back_maybe_ttl(&self) -> Option<TimeDiff> {
+        if self.sync_to_historical {
+            None
+        } else {
+            Some(self.chainspec.deploy_config.max_ttl)
+        }
+    }
+
+    fn sync_back_instruction(
+        &mut self,
+        block_synchronizer_progress: &BlockSynchronizerProgress,
+    ) -> Result<SyncBackInstruction, String> {
+        if matches!(
+            block_synchronizer_progress,
+            BlockSynchronizerProgress::Syncing(_, _, _)
+        ) {
+            return Ok(SyncBackInstruction::Syncing);
+        }
+
+        if let Some(block_header) = self.storage.get_highest_orphaned_block_header() {
+            if block_header.is_genesis() {
+                return Ok(SyncBackInstruction::GenesisSynced);
+            }
+            if let Some(diff) = self.sync_back_maybe_ttl() {
+                let cutoff = Timestamp::now().saturating_sub(diff);
```

Comment:
We shouldn't use the current timestamp, but the start time of the current era.
Of course we don't necessarily know that yet; maybe the timestamp of the most recent switch block we know of? And if we don't know any yet, don't start syncing backwards?

Reply:
lets discuss our overall approach to consistent ttl calculation and address on a separate PR

Reply:
Actually, looks like @rafal-ch addressed this side of it in a PR to my branch, which I've merged.